File: python/solver.py
from typing import Set

from board import *


class State:
    def __init__(self, board: Board):
        def all_set(n: int) -> Set:
            return set(range(1, n+1))

        self.kernel_size = board.kernel_size
        self.board_size = board.board_size
        self.row_sets = [all_set(self.board_size) for _ in range(self.board_size)]
        self.col_sets = [all_set(self.board_size) for _ in range(self.board_size)]
        self.sqr_sets = [all_set(self.board_size) for _ in range(self.board_size)]
        self.cells = np.copy(board.cells)
        self.open = 0
        for row in range(self.board_size):
            for col in range(self.board_size):
                value = board.cells[row, col]
                if value == 0:
                    self.open += 1
                else:
                    self.row_sets[row].remove(value)
                    self.col_sets[col].remove(value)
                    self.sqr_sets[self._sq_idx(row, col)].remove(value)

    def _sq_idx(self, row: int, col: int) -> int:
        return int(row/self.kernel_size)*self.kernel_size + int(col/self.kernel_size)

    def to_board(self) -> Board:
        result = Board(self.kernel_size)
        result.cells = self.cells
        return result

    def solve(self, min_hint=0, reverse=False) -> bool:
        if self.open == 0:
            return True
        else:
            row, col, open = self._min_choice_cell(min_hint)
            self.open -= 1
            for value in reversed(list(open)) if reverse else open:
                self._set(row, col, value)
                if self.solve(min_hint=len(open)-1, reverse=reverse):
                    return True
                self._unset(row, col, value)
            self.open += 1
            return False

    def _set(self, row: int, col: int, value: int):
        self.cells[row, col] = value
        self.row_sets[row].remove(value)
        self.col_sets[col].remove(value)
        self.sqr_sets[self._sq_idx(row, col)].remove(value)

    def _unset(self, row: int, col: int, value: int):
        self.cells[row, col] = 0
        self.row_sets[row].add(value)
        self.col_sets[col].add(value)
        self.sqr_sets[self._sq_idx(row, col)].add(value)

    def _min_choice_cell(self, min_hint: int):
        min_found = self.board_size
        best_row = None
        best_col = None
        best_open_set = None
        for row in range(self.board_size):
            row_set = self.row_sets[row]
            for col in range(self.board_size):
                if self.cells[row, col] == 0:
                    col_set = self.col_sets[col]
                    sq_set = self.sqr_sets[self._sq_idx(row, col)]
                    open = row_set.intersection(col_set, sq_set)
                    if len(open) < min_found:
                        min_found = len(open)
                        best_open_set = open
                        best_row = row
                        best_col = col
                        if min_found <= min_hint:
                            return best_row, best_col, best_open_set
        return best_row, best_col, best_open_set

File: sudokuNode/solver.js
const { Board } = require('./board.js');

function _range(n, start=0) {
    return Array.from({length: n}).map((_, i) => i + start);
}

function _intersect(set1, set2) {
    if (set1.length > set2.length) {
        var set3 = set1;
        set1 = set2;
        set2 = set3;
    }
    var result = new Set();
    set1.forEach(v => {
        if (set2.has(v)) {
            result.add(v)
        }
    });
    return result;
}

class State {
    constructor(board) {
        function all_set(n) {
            return new Set(_range(n, 1));
        }

        this.kernel_size = board.kernel_size;
        this.board_size = board.board_size;
        this.row_sets = _range(this.board_size).map(_ => all_set(this.board_size));
        this.col_sets = _range(this.board_size).map(_ => all_set(this.board_size));
        this.sqr_sets = _range(this.board_size).map(_ => all_set(this.board_size));
        this.cells = _range(this.board_size).map((_, i) => board.cells[i].slice());
        this.open = 0;
        _range(this.board_size).map((_, row) => {
            _range(this.board_size).map((_, col) => {
                const value = board.cells[row][col];
                if (value == 0) {
                    this.open += 1;
                }
                else {
                    this.row_sets[row].delete(value);
                    this.col_sets[col].delete(value);
                    this.sqr_sets[this._sq_idx(row, col)].delete(value);
                }
            });
        });
    }

    _sq_idx(row, col) {
        return Math.floor(row/this.kernel_size)*this.kernel_size + Math.floor(col/this.kernel_size)
    }

    to_board() {
        result = new Board(this.kernel_size);
        result.cells = this.cells
        return result
    }

    solve(min_hint=0, reverse=false) {
        const self = this;

        if (this.open == 0) {
            return true;
        }
        else {
            var [row, col, open] = this._min_choice_cell(min_hint);
            this.open -= 1;

            function try_choice(choice) {
                self._set(row, col, choice);
                if (self.solve(min_hint=open.size-1, reverse=reverse)) {
                    return true;
                }
                else {
                    self._unset(row, col, choice);
                    return false;
                }
            }

            const open_ordered = reverse ? reversed(Array.from(open)) : Array.from(open);
            const solved = open_ordered.some(try_choice);
            if (solved) {
                return true;
            }
            else {
                this.open += 1;
                return false;
            }
        }
    }

    _set(row, col, value) {
        this.cells[row][col] = value;
        this.row_sets[row].delete(value);
        this.col_sets[col].delete(value);
        this.sqr_sets[this._sq_idx(row, col)].delete(value);
    }

    _unset(row, col, value) {
        this.cells[row][col] = 0;
        this.row_sets[row].add(value);
        this.col_sets[col].add(value);
        this.sqr_sets[this._sq_idx(row, col)].add(value);
    }

    _min_choice_cell(min_hint) {
        var min_found = this.board_size;
        var best_row = null;
        var best_col = null;
        var best_open_set = null;
        _range(this.board_size).map((_, row) => {
            var row_set = this.row_sets[row];
            _range(this.board_size).map((_, col) => {
                if (this.cells[row][col] == 0) {
                    var col_set = this.col_sets[col];
                    var sq_set = this.sqr_sets[this._sq_idx(row, col)];
                    var open = _intersect(row_set, _intersect(col_set, sq_set));
                    if (open.size < min_found) {
                        min_found = open.size;
                        best_open_set = open;
                        best_row = row;
                        best_col = col;
                        if (min_found <= min_hint) {
                            return [best_row, best_col, best_open_set]
                        }
                    }
                }
            });
        });

        return [best_row, best_col, best_open_set]
    }
}

module.exports = { State }